### DISCUSSION TITLE
Help, how to manage different dependencies depending on Python version with PDM

### DISCUSSION THREAD

**Discussion by zEdS15B3GCwq** [pdm-project/pdm#1657] [Q&A]
Hi, I'm new to PDM. Trying to figure out how to manage a library project that should work with `Python >=3.8` and has a dependency that's exclusive for certain versions. More specifically, the dependency `tobii-research 1.10.2` supports **only** Python 3.10, whereas version `1.10.1` **only** supports Python 3.8.

Python 3.8
--> tobii-research 1.10.1

Python 3.10
--> tobii-research 1.10.2

I know from the docs how to create separate virtual environments with `pdm venv create` and how to switch the currently used Python version with `pdm use`, but I cannot figure out how to add one version of tobii-research in one, and another version in the other. The `pyproject.toml` and lock files can't contain multiple versions of the same package, can they? So do I need multiple versions of these?

Another issue is that PDM cannot seem to identify which version of tobii-research to use for Python 3.8. It only finds out that 1.10.2 is incompatible with Python 3.8, and instead of stepping down a version to 1.10.1, it gives up. I can manually direct it to use tobii-research==1.10.1, but I was hoping that a dependency manager could find the right version automatically. Am I perhaps using PDM wrong?

**Comment by frostming** [comment]
This is an unresolved problem #46 . It is long-hanging since it needs non-trivial effort(possibly rewrite) on the resolver.

**Reply by mforbes** [reply]
For this particular issue, would it not be possible to (perhaps with a flag) run the current dependency resolution algorithm through the allowed versions of python as if they were fixed, then incrementally add the appropriate dependencies?

For example:
```
pdm init -n
# Is there a better way of doing this?
sed 's/requires-python *= *"[^"]*"/requires-python = "~=3.6.0"/g' pyproject.toml | sponge pyproject.toml
pdm add --no-sync scipy
```

This finds, correctly, `scipy>=1.5.4`.  Repeating with python 3.7 finds `scipy>=1.7.3`, python 3.8 finds `scipt>=1.10.1` etc.  Although slow, it seems like it would be pretty trivial to have PDM do this loop, at least to establish the broad python requirements.

**Edit:** This is actually quite painful given [issue #46](https://github.com/pdm-project/pdm/issues/46#issuecomment-1482103495) since one can't use markers while increasing the python version.

**Somewhat Related Issue**:
The previous actions need to be able to find python3.6 and python3.7.  Sometimes this is hard (t is difficult to install python3.6 on new Macs with ARM).  Is there some way to run the resolver without actually trying to install the packages (i.e. cross-resolving analogous to cross-compiling)?

I.e Is there a way to properly test and/or provision packages in `pyproject.toml` cross-platform (or cross-python)?
